R/functions.R

```diff
@@ -424,3 +424,388 @@ getEligibleAtlasExperiment <- function( experiment_list, valid_experiments = eli
+        results <- list()
+
+        # Download and read TPM file if requested or in "both" mode.
+        message(paste("Downloading expression file from:\n", expressionUrl))
+        results <- tryCatch({
+            read.table(expressionUrl, header = TRUE, sep = "\t", stringsAsFactors = FALSE, na.strings = "", comment.char = "#")
+        }, error = function(e) {
+            warning(paste("Failed to download or read expression file for", experimentAccession, ":", e$message))
+            NULL
+        })
+
+        # Remove NULL results for files that could not be downloaded.
+        results <- Filter(Negate(is.null), results)
+
+        # Check if any files were successfully downloaded.
+        if (length(results) == 0) {
+            stop("ERROR - Could not download the requested data. Please check the experiment type is RNAseq baseline or try again later.")
+        }
+
+        # Message indicating success.
+        if ("expression" %in% names(results)) {
+            message(paste("Successfully downloaded expression data for", experimentAccession))
+        }
+
+        # TPMs and FPKMs can have multiple values in expression comma-separated columns (min, lower quartile, mean, higher quartile and max)
+        get_mean_value <- function(col) {
+            # Split by commas and return the third value
+            sapply(strsplit(col, ","), function(x) x[3])
+        }
+
+        # the first two columns are GeneID and Gene.Name, fetching 3rd column as it contains mean values. 
+        results[ ,3:ncol(results)] <- lapply(results[ ,3:ncol(results)], get_mean_value)
+
+        # Define filenames for configuration files.
+        configFile <- paste(experimentAccession, "-configuration.xml", sep = "")
+
+        # Define ftpUrl for configuration file.
+        configUrl <- paste(urlBase, experimentAccession, configFile, sep = "/")
+
+        # Read the XML file directly from the FTP URL
+        message("Downloading XML file from FTP...")
+        xmlContent <- tryCatch({
+            read_xml(configUrl)
+        }, error = function(e) {
+            stop("Failed to download or read the XML file: ", e$message)
+        })
+
+        # Extract the assay_group id and label attributes
+        assay_groups <- xml_find_all(xmlContent, ".//assay_group")
+
+        # Create a named vector with id as names and label as values
+        assay_vector <- setNames(xml_attr(assay_groups, "label"), xml_attr(assay_groups, "id"))
+
+        # Replace the columns names in results, excluding 'GeneID' and 'Gene.Name' columns
+        colnames(results)[-c(1, 2)] <- assay_vector[colnames(results)[-c(1, 2)]]
+
+        return(results)
+
+    } else if (normalisation %in% c("cpm")) {
+        # Define filenames for TPM and FPKM files.
+
+        experiment_summary <- getAtlasData( c(experimentAccession))[[experimentAccession]]
+
+        # Get counts.
+        results_array <- assay(experiment_summary$rnaseq, "counts")
```

Comment:
@anilthanki  These are counts not CPM, recalculation needed

Reply:
done in [ec71b97](https://github.com/ebi-gene-expression-group/bioconductor-ExpressionAtlas/pull/12/commits/ec71b973518f36447bf915b67261fb498ad4802e)